ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2151,6 +2159,173 @@ prop_connection_manager_valid_transition_order serverAcc (ArbDataFlow dataFlow)
+      evsUS :: [ObservableNetworkState SimAddr]
+      evsUS = selectTraceEventsDynamic trace
+
+      upperBound =
+        collapseCounters True
+        $ multiNodeScriptToCounters dataFlow l evsUS
+
+  in tabulate "ConnectionEvents" (map showCEvs l)
+    . counterexample (ppScript script)
+    . counterexample (Trace.ppTrace show show evsCMT)
+    . counterexample (intercalate "\n" $ map show evsUS)
+    . getAllProperty
+    . bifoldMap
+       ( \ case
+           MainReturn {} -> mempty
+           v             -> AllProperty
+                            $ counterexample (show v) (property False)
+       )
+       ( \ trs
+        -> case trs of
+          TrConnectionManagerCounters cmc ->
+            AllProperty
+              $ counterexample
+                  ("Upper bound is: " ++ show upperBound
+                  ++ "\n But got: " ++ show cmc)
+                  (property $ collapseCounters False cmc <= upperBound)
+          _                               ->
+            mempty
+       )
+    $ evsCMT
+  where
+    serverAddress :: SimAddr
+    serverAddress = TestAddress 0
+
+    -- We count all connections as prunable because we do not have a better way
+    -- to know what transitions will a given connection go through. We also
+    -- count every client connection as unidirectional because they do not
+    -- negotiate duplex data flow.
+    --
+    -- We do not count Incoming/Outgoing/Closing of connections since the
+    -- UniverseState is a much more reliable source of information.
+    multiNodeScriptToCounters :: DataFlow
+                              -> [ConnectionEvent Int TestAddr]
+                              -> [ObservableNetworkState SimAddr]
+                              -> ConnectionManagerCounters
+    multiNodeScriptToCounters df ces uss =
+      let ifDuplex = bool 0 1 (df == Duplex)
+          ifUni = bool 0 1 (df == Unidirectional)
+          ceCounters = foldMap' id
+                     . foldl'
+                        (\ st ce -> case ce of
+                          StartClient _ ta ->
+                            Map.insert ta (ConnectionManagerCounters 1 0 1 0 0) st
+                          StartServer _ ta _ ->
+                            Map.insert ta (ConnectionManagerCounters 1 ifDuplex ifUni 0 0) st
+                          _ -> st
+                        )
+                        Map.empty
+                     $ ces
+          usCounters = foldl'
+                        (\cmc (ObservableNetworkState conns) ->
+                          maxCounters cmc $
+                          Map.foldl'
+                           (\cmc' provenance ->
+                             cmc' <>
+                             if provenance == serverAddress
+                                then ConnectionManagerCounters 0 0 0 0 1
+                                else ConnectionManagerCounters 0 0 0 1 0
+                           )
+                           mempty
+                           conns
```

Comment:
It would be good to add a note that this calculation is right for the main node, because the simulation never attempts to make other connections that to or from the main node.

Reply:
This also means that we can simplify the `ceCounters`, we don't need to account counters for other nodes than the main node, isn't it?

Reply:
Yes I will add a note explaining, thanks! :D

> This also means that we can simplify the ceCounters, we don't need to account counters for other nodes than the main node, isn't it?

Every `StartClient` and `StartServer` will be associated to  an Inbound or Outbound connection request to/from the main node. Since we do not know if it will perform both (making a duplex connection), `ceCounters` look at the `StartClient` and `StartServer` events and count the number of unidirectional/duplex/prunable connections assuming it reaches the last state of the state machine (`DuplexSt`). Then the `ObservableNetworkState` is used to make a more accurate guess of the actual upper bound of inbound/outbound connections.

I'll also leave a comment explaining this.